Feature/clone function

mosquito opened the pull request:
This pull request introduces a new asynchronous file cloning feature to the `aiofile` library, allowing multiple independent file-like objects to share a single file descriptor with separate offsets. This enables concurrent reading (and potentially writing) from the same file without reopening it, which can improve performance in certain scenarios. The change includes a new `clone` helper, updates to the core `AIOFile` class, and corresponding tests and documentation.

**New file cloning feature:**

* Added a `clone` helper function and `FileIOCloner` class to `aiofile.utils`, allowing users to create independent file-like clones from an existing file object. This enables concurrent operations with separate offsets on the same file descriptor.
* Updated the public API in `aiofile/__init__.py` to export `clone` and `FileIOCloner`. [[1]](diffhunk://#diff-72ab4059288ee91d7149fdb6fcb38c696f50a93bee06cdc8aebf1b236d673e84L3-R4) [[2]](diffhunk://#diff-72ab4059288ee91d7149fdb6fcb38c696f50a93bee06cdc8aebf1b236d673e84R15) [[3]](diffhunk://#diff-72ab4059288ee91d7149fdb6fcb38c696f50a93bee06cdc8aebf1b236d673e84R25)

**Core class enhancements:**

* Modified the `AIOFile` class to support reference counting for clones, ensuring the file is only closed when all clones are released. Added a `clone()` async method and internal locking for thread safety. [[1]](diffhunk://#diff-215f902376b66188ad57674775fe016e576ca950e57e1fd24064431cdd2adbdcR142-R143) [[2]](diffhunk://#diff-215f902376b66188ad57674775fe016e576ca950e57e1fd24064431cdd2adbdcR193-R208)

**Documentation:**

* Added a new section to `README.md` explaining the `clone` helper, including a practical example and caveats for Windows users.

**Testing:**

* Added new tests in `tests/test_aio.py` to verify the behavior of file cloning, including correct reading, offset handling, and reference-counted closing. [[1]](diffhunk://#diff-614def4a76e1cf37e5de137c397dc920548ca7560707013fc38f3aea84603652L16-R16) [[2]](diffhunk://#diff-614def4a76e1cf37e5de137c397dc920548ca7560707013fc38f3aea84603652R632-R678)

Alviner commented on aiofile/utils.py:
```diff
@@ -343,13 +349,52 @@ def async_open(
+                return self.cloned_afp
+
+            self.cloned_afp = self.source_afp.__class__(
+                await self.source_afp.file.clone(),
```
What do you think about clone method in source-afp class? It would be more readable and easier to refactor in feature

mosquito committed:
Add clone helper for sharing a file descriptor across concurrent readers

AIOFile.clone() bumps a ref-count so close() is deferred until all clones
are released. FileIOCloner wraps this for the high-level async_open wrappers.
